backend/mem/fileSystem.go

```diff
@@ -0,0 +1,178 @@
+}
+
+// Retry will return a retrier provided via options, or a no-op if none is provided.
+func (fs *FileSystem) Retry() vfs.Retry {
+	return vfs.DefaultRetryer()
+}
+
+
+//NewFile function returns the in-memory implementation of vfs.File.
+//Since this is inside FileSystem, we assume that the caller knows that the CWD is the root.
+//If a non-absolute path is given, an error is thrown. Additionally, a file does not
+//technically exist until a call to "Touch()" is made on it. The "Touch" call links the
+//file with FileSystem's map and brings it into existence.
+//If a file is written to before a touch call, Write() will take care of that call.  This is
+//true for other functions as well and existence only poses a problem in the context of deletion
+//or copying FROM a non-existent file.
+func (fs *FileSystem) NewFile(volume string, absFilePath string) (vfs.File, error) {
+
+	err := utils.ValidateAbsoluteFilePath(absFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	location, err := fs.NewLocation(volume, utils.EnsureTrailingSlash(path.Dir(absFilePath)))
+	if err != nil {
+		return nil, err
+	}
+	mapRef := location.(*Location).fileSystem.fsMap
+	if _, ok := mapRef[volume];ok{
+		fileList := mapRef[volume].filesHere(location.Path())
+		for _,file := range fileList{
+			if file.name == path.Base(absFilePath){
+				fileCopy := deepCopy(file)
+				file.location = location
+				fileCopy.(*File).memFile = file
+				return fileCopy,nil
+			}
+		}
+	}
+	//validateAbsFile path will throw an error if there was a trailing slash, hence not calling path.Clean()
+	file := &File{
+		name: path.Base(absFilePath),
+	}
+
+	memFile := newMemFile(file,location)
+	file.memFile = memFile
+	return file, nil
+}
+
+
+//NewLocation function returns the in-memory implementation of vfs.Location.
+//A location always exists. If a file is created on a location that has not yet
+//been made in the fsMap, then the location will be created with the file
+func (fs *FileSystem) NewLocation(volume string, absLocPath string) (vfs.Location, error) {
+
+	err := utils.ValidateAbsoluteLocationPath(absLocPath)
+	if err != nil {
+		return nil, err
+	}
+	str := utils.EnsureTrailingSlash(path.Clean(absLocPath))
+	return &Location{
+		fileSystem: fs,
+		name:       str,
+		exists:     false,
+		volume:     volume,
+	}, nil
+
+}
+
+// Name returns the name of the underlying FileSystem
+func (fs *FileSystem) Name() string {
+	return name
+}
+
+// Scheme returns the scheme of the underlying FileSystem
+func (fs *FileSystem) Scheme() string {
+	return Scheme
+}
+
+
+//Initialize is used to initialize the fsMap for an in-memory FileSystem.
+//DISCLAIMER: nothing will work until this call is made
+
+func (fs *FileSystem) Initialize() {
+
+	fs.fsMap = make(map[string]objMap)
+
+}
+
+func init() {
+	backend.Register(Scheme, &FileSystem{sync.Mutex{},make(map[string]objMap)}) //even though the map is being made here, a call
+	// to Initialize needs to be made to properly use FileSystem
```

Comment:
This comment is strangely inline and wraps to the new line. It should be above the code block.